source/common/buffer/buffer_impl.h

```diff
@@ -1,16 +1,434 @@
+  uint64_t data_;
+
+  /** Offset in bytes from the start of the slice to the start of the Reservable section */
+  uint64_t reservable_;
+
+  /** Total number of bytes in the slice */
+  uint64_t capacity_;
+
+  /** Whether reserve() has been called without a corresponding commit(). */
+  bool reservation_outstanding_{false};
+};
+
+using SlicePtr = std::unique_ptr<Slice>;
+
+class OwnedSlice : public Slice {
+public:
+  /**
+   * Create an empty OwnedSlice.
+   * @param capacity number of bytes of space the slice should have.
+   * @return an OwnedSlice with at least the specified capacity.
+   */
+  static SlicePtr create(uint64_t capacity) {
+    uint64_t slice_capacity = sliceSize(capacity);
+    return SlicePtr(new (slice_capacity) OwnedSlice(slice_capacity));
+  }
+
+  /**
+   * Create an OwnedSlice and initialize it with a copy of the supplied copy.
+   * @param data the content to copy into the slice.
+   * @param size length of the content.
+   * @return an OwnedSlice containing a copy of the content, which may (dependent on
+   *         the internal implementation) have a nonzero amount of reservable space at the end.
+   */
+  static SlicePtr create(const void* data, uint64_t size) {
+    uint64_t slice_capacity = sliceSize(size);
+    auto slice = new (slice_capacity) OwnedSlice(slice_capacity);
+    memcpy(slice->base_, data, size);
+    slice->reservable_ = size;
+    return SlicePtr(slice);
+  }
+
+  // Custom delete operator to keep C++14 from using the global operator delete(void*, size_t),
+  // which would result in the compiler error:
+  // "exception cleanup for this placement new selects non-placement operator delete"
+  static void operator delete(void* address) { ::operator delete(address); }
+
+private:
+  static void* operator new(size_t object_size, size_t data_size) {
+    return ::operator new(object_size + data_size);
+  }
+
+  OwnedSlice(uint64_t size) : Slice(0, 0, size) { base_ = storage_; }
+
+  /**
+   * Compute a slice size big enough to hold a specified amount of data.
+   * @param data_size the minimum amount of data the slice must be able to store, in bytes.
+   * @return a recommended slice size, in bytes.
+   */
+  static uint64_t sliceSize(uint64_t data_size) {
+    static constexpr uint64_t PageSize = 4096;
+    const uint64_t num_pages = (sizeof(OwnedSlice) + data_size + PageSize - 1) / PageSize;
+    return num_pages * PageSize - sizeof(OwnedSlice);
+  }
+
+  uint8_t storage_[];
+};
+
+/**
+ * Queue of SlicePtr that supports efficient read and write access to both
+ * the front and the back of the queue.
+ * @note This class has similar properties to std::deque<T>. The reason for using
+ *       a custom deque implementation is that benchmark testing during development
+ *       revealed that std::deque was too slow to reach performance parity with the
+ *       prior evbuffer-based buffer implementation.
+ */
+class SliceDeque {
```

Comment:
this faster deque is great and is justified in comments. But I think it'd be better to move it into its own file outside of the buffer impl. and into its own header. Also github may be failing me but we need to test your new deque in isolation IMO.

My suggestion is to actually pull this into source/common/common/fast_deque.h or similar, add tests, and check that in as a separate PR. Should be easy to review and then shrink this one.

Reply:
also for tests you could start by borrowing these: https://github.com/apache/incubator-pagespeed-mod/blob/master/pagespeed/kernel/base/vector_deque_test.cc

Reply:
There's a set of tests for the new deque in `test/common/buffer/buffer_test.cc` in this PR.

Reply:
If we put this deque class in source/common/common it probably should be generic (templatized) rather than hard-coded to be a deque of slices specifically. Does the API for `SliceDeque` look reasonable as a starting point for a generic Deque class? If so, I'll convert it to a template and move it to source/common/common.

Reply:
Oh sorry I missed that you did not templatize it. If you can that'd be awesome, but that's unreasonable for me to ask in this PR. Splitting it out into its own PR seems reasonable, and putting it in its own file in the buffer directory would be good anyway so it can be individually reviewed, and easily templatized later if needed. I had missed that you added tests because ^F in github failed to find them in buffer_test.cc as it was too large to display by default, but in any case they and the class should probably be in their own file.